### PR TITLE
Digitization for hpDIRC sensor plane

### PR DESCRIPTION
**Pull request by niwgit**
### Briefly, what does this PR introduce?
This PR implements wavelength dependent quantum efficiencies for the sensor plane of hpDIRC. The "DIRC" plugin saves the hits after applying QE using "DIRCRawHits" and "DIRCRawHitsAssociations".

### What kind of change does this PR introduce?
- [ ] Bug fix (issue #__)
- [x] New feature (issue #__)
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [ ] Tests for the changes have been added
- [ ] Documentation has been added / updated
- [ ] Changes have been communicated to collaborators

### CODE REVIEW

**Comment by wdconinc** [comment]
You must enable the plugin outputs in the JEventProcessor too so they get written to the output file by default.

**Comment by wdconinc** [comment]
We should also enable optical photons in CI here

**Comment by niwgit** [comment]
> You must enable the plugin outputs in the JEventProcessor too so they get written to the output file by default.

Is this done by including "DIRCRawHits" and "DIRCRawHitsAssociations" here?
https://github.com/eic/EICrecon/blob/75e4ba32aa352d037c0c6de8fe9cf537c0917ec6/src/services/io/podio/JEventProcessorPODIO.cc#L38 ?

**Comment by wdconinc** [comment]
Yes, but no associations. We don't have associations for raw hits.

**Comment by chchatte92** [comment]
@niwgit @wdconinc just a naive question, here we are still considering the Hama MCPs; right?

**Comment by niwgit** [comment]
How do I make the "DIRC" plugin a default plugin? I don't see it on the list of default plugins.

eicrecon --list-default-plugins

 List default plugins:

  ----------------------
       Plugin name        
  ----------------------  
  log                     
  dd4hep                  
  acts                    
  richgeo                 
  rootfile                
  algorithms_calorimetry  
  algorithms_tracking     
  algorithms_digi         
  digi                    
  reco                    
  tracking                
  pid                     
  EEMC                    
  BEMC                    
  FEMC                    
  EHCAL                   
  BHCAL                   
  FHCAL                   
  B0ECAL                  
  ZDC                     
  BTRK                    
  BVTX                    
  DRICH                   
  ECTRK                   
  MPGD                    
  B0TRK                   
  RPOTS                   
  FOFFMTRK                
  BTOF                    
  ECTOF                   
  LUMISPECCAL             
  podio                   
  ----------------------

**Comment by veprbl** [comment]
> How do I make the "DIRC" plugin a default plugin? I don't see it on the list of default plugins.

In src/utilities/eicrecon/eicrecon.cc

**Comment by niwgit** [comment]
> > How do I make the "DIRC" plugin a default plugin? I don't see it on the list of default plugins.
> 
> In src/utilities/eicrecon/eicrecon.cc
Thanks!

**Comment by niwgit** [comment]
> @niwgit @wdconinc just a naive question, here we are still considering the Hama MCPs; right?

@chchatte92,
Do you mean Hamamatsu? 
If so then not exactly, in all our simulations and performance evaluations we are using PHOTONIS XP85122-S tubes (similar to Photek MAPMT 253) in case of the size of the tube and pixel size. All the parameters like QE etc. come from PANDA Barrel DIRC group measurements for larger pixel Photonis MCP-PMTs (they should be comparable though).

We did recently also simulations with Incom HRPPDs in case of the size of the tubes and pixel size, but still, all efficiencies were taken from our Photonis database (since HRPPDs don't have the values yet but promise to match the commercial tubes performance).

File: src/detectors/DIRC/DIRC.cc [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+      6.2,  6.0,  5.8,  5.7,  5.6,  5.4,  5.2,  5.1,  4.9,  4.8,  4.6,  4.5,  4.4,  4.2,  4.1,
+      4.0,  3.8,  3.7,  3.5,  3.3,  3.2,  3.1,  3.0,  2.9,  2.5,  2.4,  2.4,  2.3,  2.3,  2.1,
+      1.8,  1.6,  1.5,  1.5,  1.6,  1.8,  1.9,  1.4,  0.8,  0.9,  0.8,  0.7,  0.6,  0.3,  0.3,
+      0.5,  0.3,  0.4,  0.3,  0.1,  0.2,  0.1,  0.2,  0.3,  0.0};
+
+    for(int i=0; i < 250; i++)
+      {
+	wavelength[i] = 180 + i * 2; // wavelength units are [nm]
+	digi_cfg.quantumEfficiency.push_back({wavelength[i], sensor_qe[i]*0.01});
+      }
```
Comment (wdconinc) [inline src/detectors/DIRC/DIRC.cc:65]: Avoid multiple hard-coded 250s.
```suggestion
    std::vector<double> sensor_qe{
      0,    0,    14.0, 14.8, 14.5, 14.9, 14.4, 14.2, 13.9, 14.6, 15.2, 15.7, 16.4, 16.9, 17.5,
      17.7, 18.1, 18.8, 19.3, 19.8, 20.6, 21.4, 22.4, 23.1, 23.6, 24.1, 24.2, 24.6, 24.8, 25.2,
      25.7, 26.5, 27.1, 28.2, 29.0, 29.9, 30.8, 31.1, 31.7, 31.8, 31.6, 31.5, 31.5, 31.3, 31.0,
      30.8, 30.8, 30.4, 30.2, 30.3, 30.2, 30.1, 30.1, 30.1, 29.8, 29.9, 29.8, 29.7, 29.7, 29.7,
      29.8, 29.8, 29.9, 29.9, 29.8, 29.9, 29.8, 29.9, 29.8, 29.7, 29.8, 29.7, 29.8, 29.6, 29.5,
      29.7, 29.7, 29.8, 30.1, 30.4, 31.0, 31.3, 31.5, 31.8, 31.8, 31.9, 32.0, 32.0, 32.0, 32.0,
      32.2, 32.2, 32.1, 31.8, 31.8, 31.8, 31.7, 31.6, 31.6, 31.7, 31.5, 31.5, 31.4, 31.3, 31.3,
      31.2, 30.8, 30.7, 30.5, 30.3, 29.9, 29.5, 29.3, 29.2, 28.6, 28.2, 27.9, 27.8, 27.3, 27.0,
      26.6, 26.1, 25.9, 25.5, 25.0, 24.6, 24.2, 23.8, 23.4, 23.0, 22.7, 22.4, 21.9, 21.4, 21.2,
      20.7, 20.3, 19.8, 19.6, 19.3, 18.9, 18.7, 18.3, 17.9, 17.8, 17.8, 16.7, 16.5, 16.4, 16.0,
      15.6, 15.6, 15.2, 14.9, 14.6, 14.4, 14.1, 13.8, 13.6, 13.3, 13.0, 12.8, 12.6, 12.3, 12.0,
      11.9, 11.7, 11.5, 11.2, 11.1, 10.9, 10.7, 10.4, 10.3, 9.9,  9.8,  9.6,  9.3,  9.1,  9.0,
      8.8,  8.5,  8.3,  8.3,  8.2,  7.9,  7.8,  7.7,  7.5,  7.3,  7.1,  6.9,  6.7,  6.6,  6.3,
      6.2,  6.0,  5.8,  5.7,  5.6,  5.4,  5.2,  5.1,  4.9,  4.8,  4.6,  4.5,  4.4,  4.2,  4.1,
      4.0,  3.8,  3.7,  3.5,  3.3,  3.2,  3.1,  3.0,  2.9,  2.5,  2.4,  2.4,  2.3,  2.3,  2.1,
      1.8,  1.6,  1.5,  1.5,  1.6,  1.8,  1.9,  1.4,  0.8,  0.9,  0.8,  0.7,  0.6,  0.3,  0.3,
      0.5,  0.3,  0.4,  0.3,  0.1,  0.2,  0.1,  0.2,  0.3,  0.0};

    for (size_t i=0; i < sensor_qe.size; i++)
      {
	double wavelength = 180 + i * 2; // wavelength units are [nm]
	digi_cfg.quantumEfficiency.push_back({wavelength, sensor_qe[i]*0.01});
      }
```

File: src/detectors/DIRC/DIRC.cc [hunk 2]
```diff
@@ -0,0 +1,78 @@
+// Copyright 2022,2023 Christopher Dilks, Nilanga Wickramaarachchi
+// Subject to the terms in the LICENSE file found in the top-level directory.
+//
+//
+
+#include <JANA/JApplication.h>
+#include <JANA/JFactoryGenerator.h>
+#include <extensions/jana/JChainFactoryGeneratorT.h>
```
Comment (wdconinc) [inline src/detectors/DIRC/DIRC.cc:8]: Unused.
```suggestion
```

File: src/detectors/DIRC/DIRC.cc [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+#include <extensions/jana/JChainMultifactoryGeneratorT.h>
```
Comment (wdconinc) [inline src/detectors/DIRC/DIRC.cc:9]: Project header.
```suggestion
#include "extensions/jana/JChainMultifactoryGeneratorT.h"
```

File: src/detectors/DIRC/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+# The macro grabs sources as *.cc *.cpp *.c and headers as *.h *.hh *.hpp
+# Then correctly sets sources for ${_name}_plugin and ${_name}_library targets
+# Adds headers to the correct installation directory
+plugin_glob_all(${PLUGIN_NAME})
+
+# Find dependencies
+#plugin_add_irt(${PLUGIN_NAME})
+plugin_add_event_model(${PLUGIN_NAME})
+plugin_add_dd4hep(${PLUGIN_NAME})
+plugin_add_acts(${PLUGIN_NAME})
```
Comment (wdconinc) [inline src/detectors/DIRC/CMakeLists.txt:18]: Unused.
```suggestion
```

**Review by wdconinc — CHANGES_REQUESTED** [review]
After these changes, I think we can merge this.

**Comment by niwgit** [comment]
@wdconinc, thanks for the suggestions. I made those changes.

File: src/detectors/DIRC/DIRC.cc [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+// Subject to the terms in the LICENSE file found in the top-level directory.
+//
+//
+
+#include <JANA/JApplication.h>
+#include <JANA/JFactoryGenerator.h>
+#include "extensions/jana/JChainMultifactoryGeneratorT.h"
+
+// factories
+#include <global/digi/PhotoMultiplierHitDigi_factory.h>
```
Comment (wdconinc) [inline src/detectors/DIRC/DIRC.cc:11]:
```suggestion
#include "global/digi/PhotoMultiplierHitDigi_factory.h"
```

File: src/detectors/DIRC/DIRC.cc [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+// algorithm configurations
+#include <algorithms/digi/PhotoMultiplierHitDigiConfig.h>
```
Comment (wdconinc) [inline src/detectors/DIRC/DIRC.cc:14]:
```suggestion
#include "algorithms/digi/PhotoMultiplierHitDigiConfig.h"
```

**Review by wdconinc — APPROVED** [review]

**Pull request merged**